### PR TITLE
Ads enabled by default for Japan on upgrade

### PR DESCRIPTION
**Pull request by tmancey**
Fixes https://github.com/brave/brave-browser/issues/6449

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Test Plan:

- Confirm ads are not enabled by default for new regions
- Confirm onboarding is shown for new regions

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/brave_ads/browser/ads_service_impl.cc [hunk 1]
```diff
@@ -1318,6 +1322,21 @@ void AdsServiceImpl::MigratePrefsVersion3To4() {
   auto locale = GetLocale();
   auto region = ads::Ads::GetRegion(locale);
 
+  // Disable ads for unsupported legacy regions due to a bug where ads were
+  // enabled even if the users region was not supported
+  std::vector<std::string> legacy_regions = {
```
Comment (emerick) [inline components/brave_ads/browser/ads_service_impl.cc]: Can this be made const?

File: components/brave_ads/browser/ads_service_impl.cc [hunk 2]
```diff
@@ -1351,6 +1370,43 @@ void AdsServiceImpl::MigratePrefsVersion4To5() {
...
```
Comment (emerick) [inline components/brave_ads/browser/ads_service_impl.cc]: Can this be made const?

File: components/brave_ads/browser/ads_service_impl.cc (hunk 1 above)
Comment (tmancey) [reply components/brave_ads/browser/ads_service_impl.cc]: Fixed

File: components/brave_ads/browser/ads_service_impl.cc (hunk 2 above)
Comment (tmancey) [reply components/brave_ads/browser/ads_service_impl.cc]: Fixed

File: components/brave_ads/browser/ads_service_impl.cc [hunk 3]
```diff
@@ -1315,11 +1315,26 @@ void AdsServiceImpl::MigratePrefsVersion2To3() {
...
 
 void AdsServiceImpl::MigratePrefsVersion3To4() {
-  auto locale = GetLocale();
-  auto region = ads::Ads::GetRegion(locale);
+  const auto locale = GetLocale();
+  const auto region = ads::Ads::GetRegion(locale);
+
+  // Disable ads for unsupported legacy regions due to a bug where ads were
+  // enabled even if the users region was not supported
+  const std::vector<std::string> legacy_regions = {
```
Comment (masparrow) [inline components/brave_ads/browser/ads_service_impl.cc]: Presumably the bug has proven unfixable/untenable before?

Comment (tmancey) [reply components/brave_ads/browser/ads_service_impl.cc]: The bug was caused by Ads being incorrectly enabled in older versions so for unsupported regions we disable ads, show onboarding to allow the user to decide whether they would like to enable ads

**Comment by btlechowski** [comment]
LGTM, encountered few known issues while testing.


Brave | 0.72.103 Chromium: 78.0.3904.50 (Official Build) nightly （64 ビット）
-- | --
変更履歴 | 2accdc52c79976e264cd2694df6db31d1fccd8e8-refs/branch-heads/3904@{#658}
OS | Linux

**Clean install, Japanese locale**
Verified ads were enabled when rewards were enabled
Verified viewing an ad
Verified ads viewed increased in `brave://rewards` ads section
image

**Upgrade from 0.69.x, Japanese locale**
Verified ads are turned off by default on 0.69.x
image
Verified ads are turned off after upgrade to 0.72.x
Verified notification about ads available is shown after upgrade to 0.72.x
image
Verified that ads are not displayed after upgrade to 0.72.x
Verified clicking on `Enable Ads` in notification, opens `brave://rewards`
Verified that turning on ads, show the ads 0.72.x
image
Verified viewing an ad. 
Verified ads viewed increased in `brave://rewards` ads section
Verified ad was added to ads history

Encountered https://github.com/brave/brave-browser/issues/6088 https://github.com/brave/brave-browser/issues/6207

**Clean install, US locale**
Verified ads were enabled when rewards were enabled
Verified viewing an ad
Verified ads viewed increased in `brave://rewards` ads section

**Upgrade from 0.69.x, US locale**
Verified ads are turned on by default on 0.69.x after enabling rewards
Verified ads are turned on after upgrade
Verified ads history retained after upgrade
Verified viewing an ad

Encountered https://github.com/brave/brave-browser/issues/6270#issuecomment-542450390 when ads are turned off in 0.69.x

**Clean install, Polish locale**
Verified ads were enabled when rewards were enabled
Verified viewing an ad
Verified ads viewed increased in `brave://rewards` ads section

**Upgrade from 0.69.x, Polish locale**
Verified ads are turned off by default on 0.69.x
Verified ads are turned off after upgrade to 0.72.x
Verified notification about ads available is shown after upgrade to 0.72.x
Verified that ads are not displayed after upgrade to 0.72.x
Verified clicking on `Enable Ads` in notification, opens `brave://rewards`
Verified that turning on ads, show the ads 0.72.x
Verified viewing an ad. 
Verified ads viewed increased in `brave://rewards` ads section
Verified ad was added to ads history

**Clean install, Italian locale**
Verified ads were enabled when rewards were enabled
Verified viewing an ad
Verified ads viewed increased in `brave://rewards` ads section

**Upgrade from 0.69.x, Italian locale**
Verified ads are turned off by default on 0.69.x
Verified ads are turned off after upgrade to 0.72.x
Verified notification about ads available is shown after upgrade to 0.72.x
Verified that ads are not displayed after upgrade to 0.72.x
Verified clicking on `Enable Ads` in notification, opens `brave://rewards`
Verified that turning on ads, show the ads 0.72.x
Verified viewing an ad. 
Verified ads viewed increased in `brave://rewards` ads section
Verified ad was added to ads history

**Comment by kjozwiak** [comment]
@btlechowski awesome job on verifying this fix against Nightly 👍

**Pull request merged**